src/adapters/storage.ts

```diff
@@ -74,6 +79,59 @@ export function removeCustomToken(token: Token): Token[] {
   return setCustomTokens(getCustomTokens().filter((tkn) => tkn.address !== token.address));
 }
 
+// Pending txs
+function getSerializedPendingTxs(): SerializedPendingTx[] {
+  return getStorageByKey({
+    key: constants.PENDING_TXS_KEY,
+    defaultValue: [],
+    parser: z.array(serializedPendingTxParser),
+  });
+}
+
+function setSerializedPendingTxs(
+  serializedPendingTxs: SerializedPendingTx[]
+): SerializedPendingTx[] {
+  return setStorageByKey({
+    key: constants.PENDING_TXS_KEY,
+    value: serializedPendingTxs,
+  });
+}
+
+export function getPendingTxs(env: Env): PendingTx[] {
+  const serializedPendingTxs = getStorageByKey({
+    key: constants.PENDING_TXS_KEY,
+    defaultValue: [],
+    parser: z.array(serializedPendingTxParser),
+  });
+
+  return serializedPendingTxs.map((tx) => deserializePendingTx(tx, env));
```

Comment:
When you do `pendingTxParser(env).safeParse(tx)` you are already parsing a `SerializedPendingTx`, as they are being parsed. What's the point of parsing something that we already know its type? In this case we only want a function to convert between a `SerializedPendingTx` and a `PendingTx`, but a parser doesn't apply here. It can throw of course, in the same way a parser can throw!

You can do it in a functional way or in an imperative way... `throw catch` vs `if else`. Take a look at the last commit to see the functional implementation without using a parser. You functional guys are hard to fight! hahaha

Reply:
> When you do `pendingTxParser(env).safeParse(tx)` you are already parsing a `SerializedPendingTx`, as they are being parsed in this line: `const serializedPendingTxs = getSerializedPendingTxs();`.

Absolutely! That part was beautifully implemented. We were safely parsing what we got from the local storage, i.e the serializable entity.

> What's the purpose of parsing something that we already know its type?

You're right in that we don't *have to* use a parser in this case since we can do this check manually. I proposed using a parser just because it was a shorter and simpler implementation.

> In this case we only want a function to convert between a `SerializedPendingTx` and a `PendingTx`, but a parser doesn't apply here. It can throw of course, in the same way a parser can throw!

Well, a parser would throw only if you use `parser.parse()`. If you use `parser.safeParse()` it would return an object with success true/false and the data or the error as your last implementation does. The new implementation is ok! You have removed the throw and now return the success true/false object. Basically, we're doing manually what the parser does internally. More verbose, but perfectly safe.

> You can do it in a functional way or in an imperative way... `throw catch` vs `if else`. Take a look at the last commit to see the functional implementation without needing to use a parser twice. You functional guys are hard to fight! hahaha

I don't see how this relates to functional vs. imperative. My comment was about safe vs. unsafe.

See the initial implementation behavior when the local storage contained the `from: 3` I mentioned:

![image](https://user-images.githubusercontent.com/15896177/194021746-610f0be1-9082-4419-95df-39c0780a6f4d.png)

It was breaking the app because the function `deserializePendingTx` was throwing.

When it comes to parsing the serialized TX of the localStorage we can do it:

* With 2 parsers (what I proposed)
* With 1 parser and a function that manually does the job of the second parser (current implementation)
* With 1 parser with the signature `StrictSchema<SerializedPendingTx, PendingTx>` which first takes care of parsing the primitives we get from the local storage and second takes care of inflating the `networkId`s to the actual `Chain`. And any of the steps fail, it throws with a parsing error.